src/go/cmd/metadata-server/metadata.go

```diff
@@ -42,6 +43,59 @@ const (
 	getPodByIPWait = 500 * time.Millisecond
 )
 
+// rateLimitTokenSource is a TokenSource that applies exponential backoff on
+// errors, returning the previous error if called again too soon. It doesn't
+// rate-limit on success, as it assumes it wraps a reuseTokenSource.
+type rateLimitTokenSource struct {
+	wrapped oauth2.TokenSource
+
+	mu       sync.Mutex // guards err and next
+	err      error
+	exponent int
```

Comment:
would the code be simpler if it would state the delay and maxDelay instead of the exponent and maxExponent on initialDelay?

Reply:
I don't think so - we need to bound the value of exponent to avoid overflows (my first version counted "errors" rather than "exponent" but if "errors" got too big then `initialDelay << errors` would wrap around) - or are you imagining a different structure?

Reply:
your max delay is 102,4sec. It still feels save to double it a few times without overflow. And the upper bound would be even more explicit as it wouldn't be initial * 2^exponent but a written constant.
I was considering something similar to:

type rateLimetTokenSource struct {
delay time.Duration
...
}
func newRateLimitTokenSource(ts oauth2.TokenSource) *rateLimitTokenSource {
	return &rateLimitTokenSource{wrapped: ts, delay: 100 * time.Millisecond}
}
func (s *rateLimitTokenSource) resetBackoff() {
	s.delay = 100 * time.Millisecond
	s.next = time.Time{}
}
func (s *rateLimitTokenSource) updateBackoff(err error) {
	if s.delay <= 51200 * time.Millisecond {
		s.delay *= 2
	}
	s.next = timeNow().Add(s.delay)
	s.err = err
}

Reply:
Thanks for explaining, that is indeed simpler.